src/include/network/connection_handle_factory.h

```diff
@@ -0,0 +1,79 @@
+//===----------------------------------------------------------------------===//
+//
+//                         Peloton
+//
+// connection_handle_factory.h
+//
+// Identification: src/include/network/connection_handle_factory.h
+//
+// Copyright (c) 2015-17, Carnegie Mellon University Database Group
+//
+//===----------------------------------------------------------------------===//
+
+#pragma once
+
+#include "network/connection_handle.h"
+#include "peloton_server.h"
+
+namespace peloton {
+  namespace network {
+
+/**
+ * @brief Factory class for constructing ConnectionHandles
+ * The rationale behind using a factory is that buffers are expensive to allocate and allocating new
+ * ones every time is a bottleneck for throughput.
+ */
+class ConnectionHandleFactory {
+public:
+  /**
+   * Creates or repurpose a ConnectionHandle to be run on the given handler, handling connection from conn_fd
+   * @param conn_fd Client connection fd.
+   * @param handler The handler this ConnectionHandle is assigned to
+   * @return
+   */
+  std::shared_ptr<ConnectionHandle> GetConnectionHandle(int conn_fd, ConnectionHandlerTask *handler) {
+    // TODO(tianyu): The use of a static variable here for testing purpose is VILE. Fix this in a later refactor
+    // (probably also to-do: beat up the person who wrote this)
+    PelotonServer::recent_connfd = conn_fd;
+    auto it = reusable_handles_.find(conn_fd);
+    if (it == reusable_handles_.end()) {
+      // We are not using std::make_shared here because we want to keep ConnectionHandle constructor
+      // private to avoid unintentional use.
+      auto handle = std::shared_ptr<ConnectionHandle>(
+          new ConnectionHandle(conn_fd, handler, std::make_shared<Buffer>(), std::make_shared<Buffer>()));
+      reusable_handles_[conn_fd] = handle;
+      return handle;
+    }
+
```

Comment:
I know it's not part of this PR but am having some confusions here. Are all the ConnectionHandles exist in the `reusable_handles_` pool? And Is it also keeping those ConnectionHandles whose connections are already closed but these objects are not destroyed yet so we can reuse them later?

Reply:
Yes.